rust/parser/mod.rs

```diff
@@ -49,279 +46,361 @@ use crate::{
         TypeQLUndefine, TypeQLUpdate,
     },
 };
-use antlr_rust::{
-    token::Token,
-    tree::{ParseTree, TerminalNode as ANTLRTerminalNode},
-};
 use chrono::{NaiveDate, NaiveDateTime};
-use std::rc::Rc;
+use pest::Parser;
+use pest_derive::Parser;
 
-// keep star import to not expose generated code
-use typeql_grammar::typeqlrustparser::*;
+#[derive(Parser)]
+#[grammar = "parser/typeql.pest"]
+struct TypeQLParser;
 
-type TerminalNode<'a> = ANTLRTerminalNode<'a, TypeQLRustParserContextType>;
+type SyntaxTree<'a> = pest::iterators::Pair<'a, Rule>;
+type Children<'a> = pest::iterators::Pairs<'a, Rule>;
 
-fn get_string(string: Rc<TerminalNode>) -> String {
-    unquote(&string.get_text())
+trait TreeChildren {
+    type Items;
+    fn children(self) -> Self::Items;
 }
 
-fn get_regex(string: Rc<TerminalNode>) -> String {
-    unescape_regex(&unquote(&string.get_text()))
+impl<'a> TreeChildren for SyntaxTree<'a> {
+    type Items = Children<'a>;
+    fn children(self) -> Self::Items {
+        self.into_inner()
+    }
 }
 
-fn get_long(long: Rc<TerminalNode>) -> i64 {
-    long.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[long.get_text().as_str()]).message)
+trait RuleIterator {
+    type Item;
+
+    fn omit(&mut self, rule: Rule) -> &mut Self;
+
+    fn consume(&mut self, rule: Rule) -> Self::Item;
+    fn try_consume(&mut self, rule: Rule) -> Option<Self::Item>;
+
+    fn consume_any(&mut self) -> Self::Item;
+    fn try_consume_any(&mut self) -> Option<Self::Item>;
 }
 
-fn get_double(double: Rc<TerminalNode>) -> f64 {
-    double.get_text().parse().expect(&ILLEGAL_GRAMMAR.format(&[double.get_text().as_str()]).message)
+impl<'a, T: Iterator<Item = SyntaxTree<'a>>> RuleIterator for T {
+    type Item = SyntaxTree<'a>;
+
+    fn omit(&mut self, rule: Rule) -> &mut Self {
+        self.consume(rule);
+        self
+    }
+
+    fn consume(&mut self, _rule: Rule) -> Self::Item {
+        let next = self.consume_any();
+        assert_eq!(next.as_rule(), _rule);
+        next
+    }
+
+    fn try_consume(&mut self, _rule: Rule) -> Option<Self::Item> {
+        let next = self.try_consume_any();
+        if next.is_some() {
+            assert_eq!(next.as_ref().unwrap().as_rule(), _rule);
+        }
+        next
+    }
+
+    fn consume_any(&mut self) -> Self::Item {
+        self.next().expect("attempting to consume from an empty iterator")
+    }
+
+    fn try_consume_any(&mut self) -> Option<Self::Item> {
+        self.next()
+    }
 }
```

Comment:
Helper functions for any children iterators that allow us to validate that the rules are as we expect them. 
E.g. `iter.skip(1)` turns into `iter.omit(Rule::DEFINE)` which a) doesn't assume ownership of the iterator, b) checks that the skipped subtree has the expected tag, and c) better communicates the intent.
`Pairs` / `Children` is an iterator, so every visitor down the line uses iterator methods instead of previously used key-value methods.